pontoon/base/utils.py

```diff
@@ -398,17 +386,75 @@ def get_download_content(slug, code, part):
             vcs_translation.update_from_db(entity)
 
     resource_file.save(locale)
+    return locale_path, source_path
+
+
+def get_download_content(slug, code, part):
+    """
+    Get content of the file to be downloaded.
+
+    :param str slug: Project slug.
+    :param str code: Locale code.
+    :param str part: Resource path or Subpage name.
+    """
+    from pontoon.base.models import (
+        Locale,
+        Project,
+        Resource,
+    )
+
+    project = get_object_or_404(Project, slug=slug)
+    locale = get_object_or_404(Locale, code__iexact=code)
+
+    # Download a ZIP of all files if project has less than 10 resources
+    if 1 < project.resources.count() < 10:
+        files = {}
+
+        for resource in project.resources.all():
+            locale_path, source_path = _generate_file_for_download(resource, locale)
+
+            if not locale_path:
+                return None, None
+
+            files[resource.path] = (locale_path, source_path)
+
+        s = StringIO.StringIO()
+        zf = zipfile.ZipFile(s, "w")
+
+        for fpath, value in files.iteritems():
```

Comment:
On longer term, we'll need to be able to download a ZIP for bigger projects as well. Firefox and Firefox OS contain 100s of files and obviously the current approach won't scale.

I like your proposal with using celery. I guess generating a new ZIP after submitting each translation would be an overkill. So we could combine this with cron and only run the task every N minutes in case there were any changes affecting the ZIP.

Persistent storage would also help. (Or fake persistent storage - cloning all repositories on dyno startup). That way we'll have all files locally, we'll just have to update the changed ones and ZIP them.